platform-api/che-core-api-project/src/main/java/org/eclipse/che/api/project/server/type/ProjectTypeDetectorRegistry.java

```diff
@@ -0,0 +1,45 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2015 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.project.server.type;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Registry for {@link ProjectTypeDetector}
+ *
+ * @author Roman Nikitenko
+ */
+@Singleton
+public class ProjectTypeDetectorRegistry {
+    private Set<ProjectTypeDetector> detectors = new HashSet<>();
```

Comment:
As far as i understand this class must be thread-safe, so __HashSet__ is not the the case, or if you want to use __HashSet__ you must use synchronization for registry methods. Also i think that you should document Thread-safety as well as all methods of this registry, see [example](https://github.com/codenvy/che-core/blob/4.0/platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/RuntimeWorkspaceRegistry.java)